File: api/cuda/cuGetMaxThreads.m
function [nthreads,threadsPB] = cuGetMaxThreads
%Returns a recommended values for the maximum number concurrent threads and
% threads per block.
%   [nthreads,threadsPB] = cuGetMaxThreads
%
%This function can only be called on the host.

%#codegen -args {}

nthreads = int32(1);
threadsPB = int32(1);
if ~strcmp(coder.target,'rtw'); return; end

coder.inline('always');
coder.cinclude('cuda4m.h');

[dev, ~] = cuGetDevice;
[prop, ~] = cuGetDeviceProperties(dev);

t_prop = coder.opaque('struct cudaDeviceProp *'); %#ok<NASGU>
t_prop = coder.ceval(' ', coder.rref(prop));
nthreads = m2c_get_cstruct_field(t_prop, 'multiProcessorCount', 'int32') * ...
    m2c_get_cstruct_field(t_prop, 'maxThreadsPerMultiProcessor', 'int32');
% Set threads per block to the warp size.
threadsPB = m2c_get_cstruct_field(t_prop, 'warpSize', 'int32');
m2c_rref(prop);

File: api/cuda/codegen/lib/cuGetMaxThreads/cuGetMaxThreads.c
#include "cuGetMaxThreads.h"
#include "cuda4m.h"
#include "m2c.h"

void cuGetMaxThreads(int *nthreads, int *threadsPB)
{
  int dev;
  struct cudaDeviceProp t_prop;
  struct cudaDeviceProp * b_t_prop;
  int val;
  int b_val;
  cudaGetDevice(&dev);
  cudaGetDeviceProperties(&t_prop, dev);
  b_t_prop = (&t_prop);
  val = b_t_prop->multiProcessorCount;
  b_val = b_t_prop->maxThreadsPerMultiProcessor;
  *nthreads = val * b_val;
  *threadsPB = b_t_prop->warpSize;

}

void cuGetMaxThreads_initialize(void)
{
}

void cuGetMaxThreads_terminate(void)
{
}
